contracts/0.8.25/vaults/StakingVault.sol

```diff
@@ -404,6 +349,167 @@ contract StakingVault is IStakingVault, OwnableUpgradeable {
+
+        if (msg.sender != $.nodeOperator) revert NotAuthorized("depositToBeaconChain", msg.sender);
+        if ($.beaconChainDepositsPaused) revert BeaconChainDepositsArePaused();
+        if (valuation() < $.locked) revert ValuationBelowLockedAmount();
+
+        uint256 totalAmount = 0;
+        uint256 numberOfDeposits = _deposits.length;
+        for (uint256 i = 0; i < numberOfDeposits; i++) {
+            IStakingVault.Deposit calldata deposit = _deposits[i];
+            DEPOSIT_CONTRACT.deposit{value: deposit.amount}(
+                deposit.pubkey,
+                bytes.concat(withdrawalCredentials()),
+                deposit.signature,
+                deposit.depositDataRoot
+            );
+            totalAmount += deposit.amount;
+        }
+
+        emit DepositedToBeaconChain(msg.sender, numberOfDeposits, totalAmount);
+    }
+
+    /**
+     * @notice Calculates the total withdrawal fee required for given number of validator keys
+     * @param _numberOfKeys Number of validators' public keys
+     * @return Total fee amount to pass as `msg.value` (wei)
+     * @dev    The fee is only valid for the requests made in the same block
+     */
+    function calculateValidatorWithdrawalFee(uint256 _numberOfKeys) external view returns (uint256) {
+        if (_numberOfKeys == 0) revert ZeroArgument("_numberOfKeys");
+
+        return _numberOfKeys * TriggerableWithdrawals.getWithdrawalRequestFee();
+    }
+
+    /**
+     * @notice Requests node operator to exit validators from the beacon chain
+     *         It does not directly trigger exits - node operators must monitor for these events and handle the exits manually
+     * @param _pubkeys Concatenated validator public keys, each 48 bytes long
+     */
+    function requestValidatorExit(bytes calldata _pubkeys) external onlyOwner {
+        if (_pubkeys.length == 0) revert ZeroArgument("_pubkeys");
+        if (_pubkeys.length % PUBLIC_KEY_LENGTH != 0) {
+            revert InvalidPubkeysLength();
+        }
+
+        uint256 keysCount = _pubkeys.length / PUBLIC_KEY_LENGTH;
+        if (keysCount > MAX_PUBLIC_KEYS_PER_REQUEST) revert TooManyPubkeys();
+        for (uint256 i = 0; i < keysCount; i++) {
+            emit ValidatorExitRequested(msg.sender, string(_pubkeys[i * PUBLIC_KEY_LENGTH : (i + 1) * PUBLIC_KEY_LENGTH]));
+        }
+    }
+
+    /**
+     * @notice Triggers validator withdrawals from the beacon chain using EIP-7002 triggerable exit
+     * @param _pubkeys Concatenated validators public keys, each 48 bytes long
+     * @param _amounts Amounts of ether to exit, must match the length of _pubkeys
+     * @param _refundRecipient Address to receive the fee refund, if zero, refunds go to msg.sender
+     * @dev    The caller must provide sufficient fee via msg.value to cover the withdrawal request costs
+     */
+    function triggerValidatorWithdrawal(bytes calldata _pubkeys, uint64[] calldata _amounts, address _refundRecipient) external payable {
```

Comment:
the question is:
- deposits use high-level structs
- withdrawals use low-level concat

do we need to combine them? IDK
structs make more sense maybe

Reply:
I believe this is the subject of a separate PR